File: svws-core/src/main/java/de/svws_nrw/core/utils/klausurplan/KlausurterminblockungAlgorithmusGreedy3.java
package de.svws_nrw.core.utils.klausurplan;

import java.util.Random;
import java.util.Vector;

import de.svws_nrw.core.exceptions.DeveloperNotificationException;
import jakarta.validation.constraints.NotNull;

/**
 * Die Klausuren werden rekursiv mit Backtracking auf die Schienen verteilt.
 * Pro Rekursionsschritt wird die freie Klausur gewählt, die die meisten Nachbarsfarben hat.
 * Anschließend wird die Klausur in aufsteigender Reihenfolge auf die Schienen verteilt.
 *
 * @author Benjamin A. Bartsch
 */
public final class KlausurterminblockungAlgorithmusGreedy3 extends KlausurterminblockungAlgorithmusAbstract {

	/** Die kleinste Schienenanzahl, die bisher gefunden wurde. */
	private int _minTermine;

	/** Bis zu welchem Zeitpunkt die Rekursion laufen darf. */
	private long _zeitEnde;

	/**
	 * Konstruktor.
	 *
	 * @param pRandom   Ein {@link Random}-Objekt zur Steuerung des Zufalls über einen Anfangs-Seed.
	 * @param pDynDaten Die aktuellen Blockungsdaten.
	 */
	public KlausurterminblockungAlgorithmusGreedy3(final @NotNull Random pRandom, final @NotNull KlausurterminblockungDynDaten pDynDaten) {
		super(pRandom, pDynDaten);
	}

	@Override
	public @NotNull String toString() {
		return "Termine nacheinander, Klausurgruppen mit Backtracking";
	}

	@Override
	public void berechne(final long pZeitEnde) {
		_zeitEnde = pZeitEnde;

		// Bediene dich eines anderen Algorithmus für eine Start-Lösung.
		_dynDaten.aktion_Clear_TermineNacheinander_GruppeNachGrad();
		_minTermine = _dynDaten.gibKlausurgruppenAnzahl();
		_dynDaten.aktionZustand1Speichern();

		// Suche rekursiv nach einer besseren Lösung.
		_dynDaten.aktionClear();
		berechneRekursiv();

		// Lade die beste lokale Lösung.
		_dynDaten.aktionZustand1Laden();

		// Ist sie auch besser als die beste globale Lösung?
		if (_dynDaten.gibIstBesserAlsZustand2())
			_dynDaten.aktionZustand2Speichern();
	}

	private void berechneRekursiv() {

		// Kann das Ergebnis überhaupt noch besser werden?
		if (_dynDaten.gibTerminAnzahl() > _minTermine) return;

		// Ist die Zeit abgelaufen?
		if (System.currentTimeMillis() > _zeitEnde) return;

		// Sind alle Klausuren verteilt?
		if (!_dynDaten.gibExistierenNichtverteilteKlausuren()) {
			if (_dynDaten.gibIstBesserAlsZustand1()) {
				_minTermine = _dynDaten.gibTerminAnzahl();
				_dynDaten.aktionZustand1Speichern();
			}
			return;
		}

		// Wähle eine nächste Klausurgruppe und verteile sie rekursiv.
		final @NotNull Vector<@NotNull Integer> gruppe = _dynDaten.gibKlausurgruppeMitMinimalenTerminmoeglichkeiten();

		// 1. Fall: Die Gruppe passt noch in einen vorhandenen Termin.
		for (int terminNr = 0; terminNr < _dynDaten.gibTerminAnzahl(); terminNr++) {
			if (_dynDaten.aktionSetzeKlausurgruppeInTermin(gruppe, terminNr)) {
				berechneRekursiv();
				_dynDaten.aktionEntferneKlausurgruppeAusTermin(gruppe, terminNr);
			}
		}

		// 2. Fall: Die Gruppe muss in einen neu erzeugten Termin.
		final int terminNr = _dynDaten.gibErzeugeNeuenTermin();
		if (!_dynDaten.aktionSetzeKlausurgruppeInTermin(gruppe, terminNr)) throw new DeveloperNotificationException("Ein Setzen muss hier möglich sein!");
		berechneRekursiv();
		_dynDaten.aktionEntferneKlausurgruppeAusTermin(gruppe, terminNr);
		_dynDaten.entferneLetztenTermin();
	}

}

File: svws-webclient/core/src/core/utils/gost/klausurplanung/KlausurterminblockungAlgorithmusGreedy3.ts
import { Random } from '../../../../java/util/Random';
import { KlausurterminblockungDynDaten } from '../../../../core/utils/gost/klausurplanung/KlausurterminblockungDynDaten';
import { ArrayList } from '../../../../java/util/ArrayList';
import { KlausurterminblockungAlgorithmusAbstract } from '../../../../core/utils/gost/klausurplanung/KlausurterminblockungAlgorithmusAbstract';
import { DeveloperNotificationException } from '../../../../core/exceptions/DeveloperNotificationException';
import { System } from '../../../../java/lang/System';

export class KlausurterminblockungAlgorithmusGreedy3 extends KlausurterminblockungAlgorithmusAbstract {

	/**
	 * Die kleinste Schienenanzahl, die bisher gefunden wurde.
	 */
	private _minTermine : number = 0;

	/**
	 * Bis zu welchem Zeitpunkt die Rekursion laufen darf.
	 */
	private _zeitEnde : number = 0;


	/**
	 * Konstruktor.
	 *
	 * @param pRandom   Ein {@link Random}-Objekt zur Steuerung des Zufalls über einen Anfangs-Seed.
	 * @param pDynDaten Die aktuellen Blockungsdaten.
	 */
	public constructor(pRandom : Random, pDynDaten : KlausurterminblockungDynDaten) {
		super(pRandom, pDynDaten);
	}

	public toString() : string {
		return "Termine nacheinander, Klausurgruppen mit Backtracking";
	}

	public berechne(pZeitEnde : number) : void {
		this._zeitEnde = pZeitEnde;
		this._dynDaten.aktion_Clear_TermineNacheinander_GruppeNachGrad();
		this._minTermine = this._dynDaten.gibKlausurgruppenAnzahl();
		this._dynDaten.aktionZustand1Speichern();
		this._dynDaten.aktionClear();
		this.berechneRekursiv();
		this._dynDaten.aktionZustand1Laden();
		if (this._dynDaten.gibIstBesserAlsZustand2())
			this._dynDaten.aktionZustand2Speichern();
	}

	private berechneRekursiv() : void {
		if (this._dynDaten.gibTerminAnzahl() > this._minTermine)
			return;
		if (System.currentTimeMillis() > this._zeitEnde)
			return;
		if (!this._dynDaten.gibExistierenNichtverteilteKlausuren()) {
			if (this._dynDaten.gibIstBesserAlsZustand1()) {
				this._minTermine = this._dynDaten.gibTerminAnzahl();
				this._dynDaten.aktionZustand1Speichern();
			}
			return;
		}
		const gruppe : ArrayList<number> = this._dynDaten.gibKlausurgruppeMitMinimalenTerminmoeglichkeiten();
		for (let terminNr : number = 0; terminNr < this._dynDaten.gibTerminAnzahl(); terminNr++) {
			if (this._dynDaten.aktionSetzeKlausurgruppeInTermin(gruppe, terminNr)) {
				this.berechneRekursiv();
				this._dynDaten.aktionEntferneKlausurgruppeAusTermin(gruppe, terminNr);
			}
		}
		const terminNr : number = this._dynDaten.gibErzeugeNeuenTermin();
		if (!this._dynDaten.aktionSetzeKlausurgruppeInTermin(gruppe, terminNr))
			throw new DeveloperNotificationException("Ein Setzen muss hier möglich sein!")
		this.berechneRekursiv();
		this._dynDaten.aktionEntferneKlausurgruppeAusTermin(gruppe, terminNr);
		this._dynDaten.entferneLetztenTermin();
	}

	transpilerCanonicalName(): string {
		return 'de.svws_nrw.core.utils.gost.klausurplanung.KlausurterminblockungAlgorithmusGreedy3';
	}

	isTranspiledInstanceOf(name : string): boolean {
		return ['de.svws_nrw.core.utils.gost.klausurplanung.KlausurterminblockungAlgorithmusGreedy3', 'de.svws_nrw.core.utils.gost.klausurplanung.KlausurterminblockungAlgorithmusAbstract'].includes(name);
	}

}

export function cast_de_svws_nrw_core_utils_gost_klausurplanung_KlausurterminblockungAlgorithmusGreedy3(obj : unknown) : KlausurterminblockungAlgorithmusGreedy3 {
	return obj as KlausurterminblockungAlgorithmusGreedy3;
}
